ada/language/ast.py

```diff
@@ -4544,9 +4544,17 @@ class IndexConstraint(Constraint):
     """
     constraints = Field(type=T.ConstraintList)
 
-    xref_equation = Property(
-        Entity.constraints.logic_all(lambda c: c.xref_equation)
-    )
+    @langkit_property()
+    def xref_equation():
+        typ = Var(Entity.subtype)
+        return Entity.constraints.logic_all(
+            lambda i, c:
+            c.xref_equation
+            & c.cast(T.Expr).then(
```

Comment:
Need moar comments! Why are you casting to `Expr` and then doing a `then` ? What are you doing with the other cases ? What is that `type_bind_val` doing ?

What will your code do with the following snippet ? Will it work because the range is assigned the type `Character` explicitly ?

```ada
   subtype List_AE2 is List (Character range 'A' .. 'E');
```